### PR TITLE
Enable type-safe projections

### PR DESCRIPTION
**Pull request by dominiklohmann**
###  :notebook_with_decorative_cover: Description

This adds a free `project` function to the table slice API that allows for type-safe, row-based iteration on selected columns.

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

File-by-file. Ideally we peer-review this in a call.

### CODE REVIEW

File: libvast/src/msgpack_table_slice.cpp [hunk 1]
```diff
@@ -326,8 +326,8 @@ data_view msgpack_table_slice<FlatBuffer>::at(table_slice::size_type row,
   auto view = as_bytes(*slice_.data());
   // First find the desired row...
   VAST_ASSERT(row < offset_table.size());
-  VAST_ASSERT(state_.layout.at(*state_.layout.offset_from_index(column))->type
-              == t);
+  VAST_ASSERT(congruent(
+    state_.layout.at(*state_.layout.offset_from_index(column))->type, t));
```
Comment (dominiklohmann) [inline libvast/src/msgpack_table_slice.cpp:330]: Note to the reviewer: This change was a bug fix. The old assertion needlessly triggered for aliases and types with attributes.

File: libvast/vast/project.hpp [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+          if (name_view == hint)
+            if (congruent(field.type(), type))
+              return flat_index;
+          auto colon = name_view.find_first_of('.');
+          if (colon == std::string_view::npos)
+            break;
+          name_view.remove_prefix(colon + 1);
+        }
+        ++flat_index;
+      }
```
Comment (dominiklohmann) [inline libvast/vast/project.hpp:253]: Create story for streamlining the `record_type` API; this should not be as complicated.

**Review by dominiklohmann — COMMENTED** [review]
Peer-reviewed with @tobim.

File: libvast/vast/project.hpp [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+      return std::invoke(invocable, *std::forward<decltype(args)>(args)...);
+    },
+    map_tuple_elements_get_addresses<Indices>(tuples...))...);
+}
+
+// Invoke function for each of the given tuples, capturing the result in a
+// tuple itself.
+template <class Invocable, class Tuple, class... Tuples>
+constexpr auto
+map_tuple_elements(Invocable& invocable, Tuple&& tuple, Tuples&&... tuples) {
```
Comment (dominiklohmann) [inline libvast/vast/project.hpp:46]: tuple_zip?

**Review by tobim — APPROVED** [review]
Please address my 1 comment, then feel free to merge.

File: libvast/vast/project.hpp (hunk 2 above)
Comment (dominiklohmann) [reply libvast/vast/project.hpp]: Done.

**Pull request merged**